test/e2e_node/pod_termination_state_test.go

```diff
@@ -0,0 +1,182 @@
+		ginkgo.BeforeEach(func() {
+			ginkgo.By("setting up the pod to be used in the test")
+			podClient = e2epod.NewPodClient(f)
+			podClient.Create(context.TODO(), podSpec)
+		})
+
+		ginkgo.It("The pod should be running and deleted with grace period", func() {
+
+			// We need to wait for the pod to be running, otherwise the deletion
+			// may be carried out immediately rather than gracefully.
+			framework.ExpectNoError(e2epod.WaitForPodNameRunningInNamespace(context.TODO(), f.ClientSet, podSpec.Name, f.Namespace.Name))
+
+			// save the running pod
+			pod, err := podClient.Get(context.TODO(), podSpec.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to GET scheduled pod")
+
+			// save the UID for cgroup verify
+			podUID := string(pod.UID)
+
+			// delete pod gracefully
+			ginkgo.By("deleting the pod gracefully")
+			var lastPod v1.Pod
+			var statusCode int
+			err = f.ClientSet.CoreV1().RESTClient().Delete().AbsPath("/api/v1/namespaces", pod.Namespace, "pods", pod.Name).Param("gracePeriodSeconds", "30").Do(context.TODO()).StatusCode(&statusCode).Into(&lastPod)
+			framework.ExpectNoError(err, "failed to use http client to send delete")
+			framework.ExpectEqual(statusCode, http.StatusOK, "failed to delete gracefully by client request")
+
+			// verifying the kubelet observed the termination notice
+			// check pod status is from running to terminating status, after a while, pod is deleted in kubelet
+			err = e2epod.WaitForPodTerminatingInNamespaceTimeout(context.TODO(), f.ClientSet, podSpec.Name, f.Namespace.Name, 30)
+			framework.ExpectNoError(err)
+
+			// Sleep enough time to ensure the pod is terminated
+			time.Sleep(30 * time.Second)
+
+			// resources are cleaned up and phase is set correctly cgroup files
+			// see same way in pods_container_manager_test.go
+			new_pod := makePodToVerifyCgroupRemoved("pod" + podUID)
+			podClient.Create(context.TODO(), new_pod)
+			err = e2epod.WaitForPodSuccessInNamespace(context.TODO(), f.ClientSet, new_pod.Name, f.Namespace.Name)
+			framework.ExpectNoError(err)
+
+			// verify pod is deleted from API server
+			ginkgo.By("Fetching the latest pod status")
+			err = e2epod.WaitForPodNotFoundInNamespace(context.TODO(), f.ClientSet, podSpec.Name, f.Namespace.Name, 120)
+			framework.ExpectNoError(err)
+		})
+
+		ginkgo.AfterEach(func() {
+			ginkgo.By(fmt.Sprintf("deleting pod: %s", podSpec.Name))
+			e2epod.NewPodClient(f).DeleteSync(context.TODO(), podSpec.Name, metav1.DeleteOptions{}, framework.PodDeleteTimeout)
+		})
+	})
+}
+
+func runEvictedTest(f *framework.Framework, podSpec *v1.Pod, containerName string) {
+	ginkgo.Context("", func() {
+		var podClient *e2epod.PodClient
+		ginkgo.BeforeEach(func() {
+			ginkgo.By("setting up the pod to be used in the test")
+			podClient = e2epod.NewPodClient(f)
+			podClient.Create(context.TODO(), podSpec)
+		})
+
+		ginkgo.It("The containers terminated by OOM killer should have the reason set to OOMKilled", func() {
```

Comment:
Is this a bit duplicated with the existing OOM killer test (https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/oomkiller_test.go)? Is the only difference we are also checking `makePodToVerifyCgroupRemoved`? If so, does it maybe make sense to just add that check there? (or maybe longer term, we should add this there? or ideally we could have this check for all pod terminations in node e2e?

Reply:
yeah. Previously, the test intends to check terminating status and pod deleted from API server, besides the cgroup deletion. However in the oomkill cases, it is hard to check the terminating status. and as it is failed pod, it takes long time for pod to be deleted in API server. 

That's why only check cgroup left here. Yeah, we can add `makePodToVerifyCgroupRemoved ` to OOM killer test instead of duplicating here.